### DISCUSSION TITLE
Need --dither-depth=10?

### DISCUSSION THREAD

**Discussion by rainlycoris** [iina/iina#5157] [Q&A] [answered]
> Note that the on-the-wire bit depth cannot be detected except when using gpu-api=d3d11. Explicitly setting the value to your display's bit depth is recommended, as dithering performed by some LCD panels can be of low quality.

The MacBook air M2 or MacBook pro M1 should be followed by a 10bit panel, so should you set '--dither-depth=10' to get the best quality?
Now `bufferDepth` is `VideoLayer` right? If set `dither-depth=10` will be a problem when playing 8 bit videos?

**Comment by low-batt** [comment, marked as answer]
The changes contained in PR #5154 will add a `bufferDepth` property to `VideoLayer`, assuming the proposed changes pass review. That name was chosen to match up with mpv. That PR also adds code to pass `bufferDepth` to the renderer:
```swift
mpv_render_param(type: MPV_RENDER_PARAM_DEPTH, data:.init(bufferDepth)),
```

The description of that renderer parameter is:
```c
    /**
     * Control surface depth. Valid for mpv_render_context_render().
     * Type: int*
     * This implies the depth of the surface passed to the render function in
     * bits per channel. If omitted or set to 0, the renderer will assume 8.
     * Typically used to control dithering.
     */
    MPV_RENDER_PARAM_DEPTH = 5,
```

I have not investigated the mpv code, so I could be wrong, but I think the value provided for MPV_RENDER_PARAM_DEPTH will be used as the "detected depth" mentioned in the documentation for [dither-depth](https://mpv.io/manual/stable/#options-dither-depth).

The value of `bufferDepth` is based on the mpv [cocoa-cb-10bit-context](https://mpv.io/manual/stable/#options-cocoa-cb-10bit-context) option:
> **--cocoa-cb-10bit-context=<yes|no>**
>    Creates a 10bit capable pixel format for the context creation (default: yes). Instead of 8bit integer framebuffer a 16bit half-float framebuffer is requested.

As that option defaults to `yes`, IINA when built with the changes in PR #5154 will use `10` as the `dither-depth`.

On this:
> If set `dither-depth=10` will be a problem when playing 8 bit videos?

No, it will not be a problem. The `dither-depth` is informing mpv of the capability of the display. This is needed when [dithering](https://en.wikipedia.org/wiki/Dither#Digital_photography_and_image_processing) the image. This comes into play when the video has more colors than the display can represent.

At least that is my understanding. I'm not an expert when it comes to color management.

With IINA 1.3.5 if you set `dither-depth` to `10` in advanced settings and play a 10 bit video on a 10 bit display it should improve the image. The improvement may be subtle.

**Reply by rainlycoris** [reply]
Thank you. I have one more question. If --cocoa-cb-10bit-context=no, will it result in only 8-bit pixels being output? If the dither to 10bit will not be converted to 8bit and then output the problem?

**Reply by low-batt** [reply]
Excellent question. Although [cocoa-cb-10bit-context](https://mpv.io/manual/stable/#options-cocoa-cb-10bit-context) defaults to `yes` it might as well be `no` since IINA 1.3.5 is missing the code required to support 10 bit. Support for that is being added by PR #5154.

Really need a mpv developer to comment on how 10 bit dithering would react when being applied to an 8 bit layer. I would expect it would decide there is nothing for it to do. Dithering is used when displaying an image with many colors on a display that is limited. Setting [dither-depth](https://mpv.io/manual/stable/#options-dither-depth) to 10 tells mpv that the display supports [Deep color](https://en.wikipedia.org/wiki/Color_depth#Deep_color_(30-bit)). Thus the display is plenty capable of displaying an 8 bit image without dithering.

It should be noted that 10 bit displays are usually 8 bit displays that use [FRC](https://en.wikipedia.org/wiki/Frame_rate_control). See [What’s the Difference Between an 8-Bit + FRC and a 10-Bit Monitor?](https://www.benq.com/en-us/knowledge-center/knowledge/10-bit-vs-8-bit-frc-monitor-color-difference.html) for a discussion about that.

**Reply by rainlycoris** [reply]
I think the video renders at rgba16f, examples resampling and glsl, it's always going from high precision to low precision. So I guess mpv will always dither.
At least the MacBook pro M1 and Pro Display XDR has a native 10bit panel, I didn't know MacBook air M2.

**Reply by rainlycoris** [reply]
Can you add an option to always output at 10bit in iina and set `--dither-depth=10`?
Because it seems that 8bit is still output at 8bit at the moment, and I don't know how to change `bufferDepth=10` in mpv.conf.

And the video is always rendered at rgba16f, so there must be some jitter, even for 8-bit videos.

The Macbook air M2 is also a native 10bit panel.

While 10bit output may not improve perceived quality, it should be free of charge for performance and reassuring.

**Reply by rainlycoris** [reply]
Sorry, my github was suspended and I'm only now unblocked. Because they think they're doing something suspicious, and respond my ticket back now.

So what do you think?

**Reply by low-batt** [reply]
The `bufferDepth` I mentioned has to do with internal code changes. It is not a property that can be set in the mpv conf file. Sorry about the confusion.

The mpv [--dither-depth](https://mpv.io/manual/stable/#options-dither-depth) option defaults to `auto`. With IINA 1.3.5 mpv picks `8`. With the changes in PR #5154 that add 10 bit support the mpv log shows:
```text
[  16.348][v][libmpv_render] Dither to 16.
```

Those changes still have to be reviewed and merged by the senior developers.

**Reply by rainlycoris** [reply]
That's why I'm requesting you to add this setting, because bufferDepth can't be set in mpv.conf. Just alaways set `bufferDepth=10`.
Sorry, I don't know what `bufferDepth` is, I just thought it might have an effect on 10bit output.
What the log showed, I didn't understand (

**Reply by low-batt** [reply]
Three major software components are involved:
- AppKit/OpenGL (Apple)
- `libmpv` ([mpv project](https://github.com/mpv-player/mpv))
- IINA

The majority of the work is performed by `libmpv` and AppKit/OpenGL. Think of IINA as the glue that ties these two together. Although nearly all processing of mpv options occurs in `libmpv`, some options require some support in the "glue" (`libmpv` client). For the issue at hand, the two options we care about are:
- [--cocoa-cb-10bit-context](https://mpv.io/manual/stable/#options-cocoa-cb-10bit-context)
- [--dither-depth](https://mpv.io/manual/stable/#options-dither-depth)

Both of those options require support in the `libmpv` client (the "glue"/IINA). IINA 1.3.5 is missing that support. The option that provides 10 bit support:

> **--cocoa-cb-10bit-context=<yes|no>**
>
>    Creates a 10bit capable pixel format for the context creation (default: yes). Instead of 8bit integer framebuffer a 16bit half-float framebuffer is requested.

Defaults to `yes`. So there is no need to set it to obtain 10 bit support. Trouble is it does not work with IINA 1.3.5. The [mpv player](https://mpv.io/) also uses `libmpv`, so its sources provide a reference implementation on how to properly use `libmpv`.  The mpv class `gl_layer` contains a property `bufferDepth` as seen [here](https://github.com/mpv-player/mpv/blob/53f2619dbd740b71bdd4277f096d077ffb8d571b/video/out/mac/gl_layer.swift#L80). That property is set based on this option. Part of that code can be seen [here](https://github.com/mpv-player/mpv/blob/53f2619dbd740b71bdd4277f096d077ffb8d571b/video/out/mac/gl_layer.swift#L281-L283).

This code is not in `libmpv`. This code is needed to properly use `libmpv`, so IINA needs similar code for the `-cocoa-cb-10bit-context` option to work. The equivalent code in IINA is the [ViewLayer](https://github.com/iina/iina/blob/develop/iina/ViewLayer.swift) class. This class is missing any such code. IINA 1.3.5 only supports an 8 bit pixel buffer. The changes in PR #5154 add a `bufferDepth` property and associated code similar to what mpv does to the `ViewLayer` class so that `-cocoa-cb-10bit-context` will work.

So think of `bufferDepth` as an internal implementation detail. It is the `-cocoa-cb-10bit-context` option that controls whether 8 or 10 bit is used. But IINA 1.3.5 is missing support for `-cocoa-cb-10bit-context` and `--dither-depth`. These options will not work with IINA 1.3.5.

With the proposed changes in PR #5154, `-cocoa-cb-10bit-context` works. Those changes also cause `--dither-depth=auto` to choose 16 for the depth when a 10 bit context is being used. It is 16 instead of 10 because a short (16 bit) floating point format is used.

These changes still need to be reviewed by the senior IINA developers. They have not been added to IINA as of yet. In the review I will ask about adding explicit UI controls for these settings.

Late here so my reply may be a little incoherent. If it is not making sense please do continue to ask questions. I'm happy to explain.
